### ISSUE TITLE
No handler for message "FOS\ElasticaBundle\Message\AsyncPersistPage"

### ISSUE THREAD

**Issue by darthf1** [FriendsOfSymfony/FOSElasticaBundle#1720]
Hi! I'm trying `v6.0.0-beta3` with the `Symfony Messenger (v5.1.7)` integration.

**TLDR:**
- All the messages fail directly with `No handler for message "FOS\ElasticaBundle\Message\AsyncPersistPage".
- Manually retrying with `php bin/console messenger:failed:retry 170` gives the same result.
- When I manually register the handler, the messages still fail directly.
- Manually retrying the message now "magically" works

-----------------

## First failure:
```
php bin/console messenger:failed:show 170 -vv

[2020-10-14T15:27:16.324155+02:00] doctrine.DEBUG: SELECT COUNT(m.id) as message_count FROM messenger_messages m WHERE (m.delivered_at is null OR m.delivered_at < ?) AND (m.available_at <= ?) AND (m.queue_name = ?) LIMIT 1 ["2020-10-14T14:27:16+02:00","2020-10-14T15:27:16+02:00","failed"] []
[2020-10-14T15:27:16.326446+02:00] doctrine.DEBUG: SELECT COUNT(m.id) as message_count FROM messenger_messages m WHERE (m.delivered_at is null OR m.delivered_at < ?) AND (m.available_at <= ?) AND (m.queue_name = ?) LIMIT 1 ["2020-10-14T14:27:16+02:00","2020-10-14T15:27:16+02:00","failed"] []
There are 42 messages pending in the failure transport.
[2020-10-14T15:27:16.326921+02:00] doctrine.DEBUG: SELECT m.* FROM messenger_messages m WHERE m.id = ? ["170"] []

Failed Message Details
======================

 ------------- ----------------------------------------------------------------------- 
  Class         FOS\ElasticaBundle\Message\AsyncPersistPage                            
  Message Id    170                                                                    
  Failed at     2020-10-14 15:18:03                                                    
  Error         No handler for message "FOS\ElasticaBundle\Message\AsyncPersistPage".  
  Error Class   Symfony\Component\Messenger\Exception\NoHandlerForMessageException     
  Transport     default                                                                
 ------------- ----------------------------------------------------------------------- 

 Message history:
  * Message failed at 2020-10-14 15:17:56 and was redelivered
  * Message failed at 2020-10-14 15:17:57 and was redelivered
  * Message failed at 2020-10-14 15:17:59 and was redelivered
  * Message failed at 2020-10-14 15:18:03 and was redelivered

Message:
========

FOS\ElasticaBundle\Message\AsyncPersistPage^ {
  -page: 6
  -options: [
    "max_per_page" => 100,
    "first_page" => 1,
    "last_page" => 42,
    "delete" => true,
    "reset" => true,
    "ignore_errors" => false,
    "sleep" => 0,
    "indexName" => "MyEntity"
  ]
}

Exception:
==========

#0 /srv/api/vendor/symfony/messenger/Middleware/SendMessageMiddleware.php(74): Symfony\Component\Messenger\Middleware\HandleMessageMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#1 /srv/api/vendor/symfony/messenger/Middleware/FailedMessageProcessingMiddleware.php(34): Symfony\Component\Messenger\Middleware\SendMessageMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#2 /srv/api/vendor/symfony/messenger/Middleware/DispatchAfterCurrentBusMiddleware.php(67): Symfony\Component\Messenger\Middleware\FailedMessageProcessingMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#3 /srv/api/vendor/symfony/messenger/Middleware/RejectRedeliveredMessageMiddleware.php(48): Symfony\Component\Messenger\Middleware\DispatchAfterCurrentBusMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#4 /srv/api/vendor/symfony/messenger/Middleware/AddBusNameStampMiddleware.php(37): Symfony\Component\Messenger\Middleware\RejectRedeliveredMessageMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#5 /srv/api/vendor/symfony/messenger/MessageBus.php(80): Symfony\Component\Messenger\Middleware\AddBusNameStampMiddleware->handle(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Middleware\StackMiddleware))
#6 /srv/api/vendor/symfony/messenger/TraceableMessageBus.php(41): Symfony\Component\Messenger\MessageBus->dispatch(Object(Symfony\Component\Messenger\Envelope), Array)
#7 /srv/api/vendor/symfony/messenger/RoutableMessageBus.php(54): Symfony\Component\Messenger\TraceableMessageBus->dispatch(Object(Symfony\Component\Messenger\Envelope), Array)
#8 /srv/api/vendor/symfony/messenger/Worker.php(113): Symfony\Component\Messenger\RoutableMessageBus->dispatch(Object(Symfony\Component\Messenger\Envelope))
#9 /srv/api/vendor/symfony/messenger/Worker.php(77): Symfony\Component\Messenger\Worker->handleMessage(Object(Symfony\Component\Messenger\Envelope), Object(Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport), 'default')
#10 /srv/api/vendor/symfony/messenger/Command/ConsumeMessagesCommand.php(197): Symfony\Component\Messenger\Worker->run(Array)
#11 /srv/api/vendor/symfony/console/Command/Command.php(258): Symfony\Component\Messenger\Command\ConsumeMessagesCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#12 /srv/api/vendor/symfony/console/Application.php(938): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /srv/api/vendor/symfony/framework-bundle/Console/Application.php(96): Symfony\Component\Console\Application->doRunCommand(Object(Symfony\Component\Messenger\Command\ConsumeMessagesCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 /srv/api/vendor/symfony/console/Application.php(266): Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand(Object(Symfony\Component\Messenger\Command\ConsumeMessagesCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /srv/api/vendor/symfony/framework-bundle/Console/Application.php(82): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /srv/api/vendor/symfony/console/Application.php(142): Symfony\Bundle\FrameworkBundle\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /srv/api/bin/console(43): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput))
#18 {main}

 Run messenger:failed:retry 170 to retry this message.
 Run messenger:failed:remove 170 to delete it.
```
Command to populate:
```
php bin/console fos:elastica:populate --pager-persister=async
```

`fos_elastica.yaml`
```yaml
# Read the documentation: https://github.com/FriendsOfSymfony/FOSElasticaBundle/blob/master/doc/setup.md
fos_elastica:
  messenger:
    enabled: true
  serializer: ~
  clients:
    default:
      url: '%env(ELASTICSEARCH_URL)%'
      compression: true
  indexes:
    myEntity:
      persistence:
        driver: orm
        model: App\Entity\myEntity
      serializer:
        groups: ["myEntity:read"]
```

`messenger.yaml`
```yaml
framework:
  messenger:
    default_bus: command.bus

    # Uncomment this (and the failed transport below) to send failed messages to this transport for later handling.
    failure_transport: failed
    transports:
      # https://symfony.com/doc/current/messenger.html#transport-configuration
      default:
        dsn: "%env(MESSENGER_TRANSPORT_DSN)%?queue_name=default"
        retry_strategy:
          max_retries: 3
          delay: 1000
          multiplier: 2
          max_delay: 0
      failed: "%env(MESSENGER_TRANSPORT_DSN)%?queue_name=failed"
      # sync: "sync://"

    routing:
      # Route your messages to the transports
      "*": default

    buses:
      command.bus: ~
```

`supervisor.conf`
```
[program:messenger-consume-default]
command=/srv/api/bin/console messenger:consume default --time-limit=3600 --no-debug
user=root
numprocs=2
autostart=true
autorestart=true
process_name=%(program_name)s_%(process_num)02d
stdout_logfile=/srv/api/var/log/supervisor-%(program_name)s_stdout.log
stderr_logfile_maxbytes=10MB
stderr_logfile=/srv/api/var/log/supervisor-%(program_name)s_stderr.log
stdout_logfile_maxbytes=10MB
```

When debugging messenger, it doesn't show the message nor the handler.

`php bin/console debug:messenger`
```
Messenger
=========

command.bus
-----------

 The following messages can be dispatched:

 ---------------------------------------------------------------------------------- 
  Symfony\Component\Mailer\Messenger\SendEmailMessage                               
      handled by mailer.messenger.message_handler                                   
                                                                                    
  Represents an update to send to the hub.                                          
  Symfony\Component\Mercure\Update                                                  
      handled by mercure.hub.default.publisher                               
                                                                                    
 ---------------------------------------------------------------------------------- 
```

`php bin/console debug:container | grep fos`
```
  Elastica\Client                                                                                                        alias for "fos_elastica.client.default"                                                                                        
  FOS\ElasticaBundle\Configuration\ConfigManager                                                                         alias for "fos_elastica.config_manager"                                                                                        
  FOS\ElasticaBundle\Elastica\Client                                                                                     alias for "fos_elastica.client.default"                                                                                        
  FOS\ElasticaBundle\Index\IndexManager                                                                                  alias for "fos_elastica.index_manager"                                                                                         
  FOS\ElasticaBundle\Index\IndexTemplateManager                                                                          alias for "fos_elastica.index_template_manager"                                                                                
  FOS\ElasticaBundle\Index\Resetter                                                                                      alias for "fos_elastica.resetter"                                                                                              
  FOS\ElasticaBundle\Index\TemplateResetter                                                                              alias for "fos_elastica.template_resetter"                                                                                     
  FOS\ElasticaBundle\Manager\RepositoryManagerInterface                                                                  alias for "fos_elastica.manager.orm"                                                                                           
  fos_elastica.alias_processor                                                                                           FOS\ElasticaBundle\Index\AliasProcessor                                                                                        
  fos_elastica.async_pager_persister                                                                                     FOS\ElasticaBundle\Persister\AsyncPagerPersister                                                                               
  fos_elastica.client                                                                                                    alias for "fos_elastica.client.default"                                                                                        
  fos_elastica.client.default                                                                                            FOS\ElasticaBundle\Elastica\Client                                                                                             
  fos_elastica.client_prototype                                                                                          FOS\ElasticaBundle\Elastica\Client                                                                                             
  fos_elastica.command.create                                                                                            FOS\ElasticaBundle\Command\CreateCommand                                                                                       
  fos_elastica.command.delete                                                                                            FOS\ElasticaBundle\Command\DeleteCommand                                                                                       
  fos_elastica.command.populate                                                                                          FOS\ElasticaBundle\Command\PopulateCommand                                                                                     
  fos_elastica.command.reset                                                                                             FOS\ElasticaBundle\Command\ResetCommand                                                                                        
  fos_elastica.command.search                                                                                            FOS\ElasticaBundle\Command\SearchCommand                                                                                       
  fos_elastica.command.templates_reset                                                                                   FOS\ElasticaBundle\Command\ResetTemplatesCommand                                                                               
  fos_elastica.config_manager                                                                                            FOS\ElasticaBundle\Configuration\ConfigManager                                                                                 
  fos_elastica.config_manager.index_templates                                                                            FOS\ElasticaBundle\Configuration\ConfigManager                                                                                 
  fos_elastica.config_source.container                                                                                   FOS\ElasticaBundle\Configuration\Source\ContainerSource                                                                        
  fos_elastica.config_source.template_container                                                                          FOS\ElasticaBundle\Configuration\Source\TemplateContainerSource                                                                
  fos_elastica.data_collector                                                                                            FOS\ElasticaBundle\DataCollector\ElasticaDataCollector                                                                         
  fos_elastica.doctrine.register_listeners                                                                               FOS\ElasticaBundle\Doctrine\RegisterListenersService                                                                           
  fos_elastica.elastica_to_model_transformer.myEntity                                                                    FOS\ElasticaBundle\Doctrine\ORM\ElasticaToModelTransformer                                                                     
  fos_elastica.elastica_to_model_transformer.prototype.orm                                                               FOS\ElasticaBundle\Doctrine\ORM\ElasticaToModelTransformer                                                                     
  fos_elastica.filter_objects_listener                                                                                   FOS\ElasticaBundle\Persister\Listener\FilterObjectsListener                                                                    
  fos_elastica.finder                                                                                                    FOS\ElasticaBundle\Finder\TransformedFinder                                                                                    
  fos_elastica.finder.myEntity                                                                                           FOS\ElasticaBundle\Finder\TransformedFinder                                                                                    
  fos_elastica.in_place_pager_persister                                                                                  FOS\ElasticaBundle\Persister\InPlacePagerPersister                                                                             
  fos_elastica.index                                                                                                     alias for "fos_elastica.index.myEntity"                                                                                    
  fos_elastica.index.myEntity                                                                                            FOS\ElasticaBundle\Elastica\Index                                                                                              
  fos_elastica.index.myEntity.serializer.callback                                                                        FOS\ElasticaBundle\Serializer\Callback                                                                                         
  fos_elastica.index_manager                                                                                             FOS\ElasticaBundle\Index\IndexManager                                                                                          
  fos_elastica.index_prototype                                                                                           FOS\ElasticaBundle\Elastica\Index                                                                                              
  fos_elastica.index_template_manager                                                                                    FOS\ElasticaBundle\Index\IndexTemplateManager                                                                                  
  fos_elastica.index_template_prototype                                                                                  FOS\ElasticaBundle\Elastica\IndexTemplate                                                                                      
  fos_elastica.indexable                                                                                                 FOS\ElasticaBundle\Provider\Indexable                                                                                          
  fos_elastica.listener.myEntity                                                                                         FOS\ElasticaBundle\Doctrine\Listener                                                                                           
  fos_elastica.listener.prototype.orm                                                                                    FOS\ElasticaBundle\Doctrine\Listener                                                                                           
  fos_elastica.logger                                                                                                    FOS\ElasticaBundle\Logger\ElasticaLogger                                                                                       
  fos_elastica.manager                                                                                                   alias for "fos_elastica.manager.orm"                                                                                           
  fos_elastica.manager.orm                                                                                               FOS\ElasticaBundle\Doctrine\RepositoryManager                                                                                  
  fos_elastica.mapping_builder                                                                                           FOS\ElasticaBundle\Index\MappingBuilder                                                                                        
  fos_elastica.model_to_elastica_identifier_transformer                                                                  FOS\ElasticaBundle\Transformer\ModelToElasticaIdentifierTransformer                                                            
  fos_elastica.model_to_elastica_transformer                                                                             FOS\ElasticaBundle\Transformer\ModelToElasticaAutoTransformer                                                                  
  fos_elastica.model_to_elastica_transformer.myEntity                                                                    FOS\ElasticaBundle\Transformer\ModelToElasticaIdentifierTransformer                                                            
  fos_elastica.object_persister                                                                                          FOS\ElasticaBundle\Persister\ObjectPersister                                                                                   
  fos_elastica.object_persister.myEntity                                                                                 FOS\ElasticaBundle\Persister\ObjectSerializerPersister                                                                         
  fos_elastica.object_serializer_persister                                                                               FOS\ElasticaBundle\Persister\ObjectSerializerPersister                                                                         
  fos_elastica.pager_persister_registry                                                                                  FOS\ElasticaBundle\Persister\PagerPersisterRegistry                                                                            
  fos_elastica.pager_provider.myEntity                                                                                   FOS\ElasticaBundle\Doctrine\ORMPagerProvider                                                                                   
  fos_elastica.pager_provider.prototype.orm                                                                              FOS\ElasticaBundle\Doctrine\ORMPagerProvider                                                                                   
  fos_elastica.pager_provider_registry                                                                                   FOS\ElasticaBundle\Provider\PagerProviderRegistry                                                                              
  fos_elastica.paginator.subscriber                                                                                      FOS\ElasticaBundle\Subscriber\PaginateElasticaQuerySubscriber                                                                  
  fos_elastica.persister_registry                                                                                        FOS\ElasticaBundle\Persister\PersisterRegistry                                                                                 
  fos_elastica.populate_listener                                                                                         FOS\ElasticaBundle\EventListener\PopulateListener                                                                              
  fos_elastica.property_accessor                                                                                         Symfony\Component\PropertyAccess\PropertyAccessor                                                                              
  fos_elastica.repository_manager                                                                                        FOS\ElasticaBundle\Manager\RepositoryManager                                                                                   
  fos_elastica.resetter                                                                                                  FOS\ElasticaBundle\Index\Resetter                                                                                              
  fos_elastica.serializer                                                                                                alias for "serializer"                                                                                                         
  fos_elastica.serializer_callback_prototype                                                                             FOS\ElasticaBundle\Serializer\Callback                                                                                         
  fos_elastica.template_resetter                                                                                         FOS\ElasticaBundle\Index\TemplateResetter
```

--------
## Manually registering the services

I tried registering the handler manually:
`services.yaml`:
```yaml
  FOS\ElasticaBundle\Persister\AsyncPagerPersister:
    arguments:
      $pagerPersisterRegistry: '@fos_elastica.pager_persister_registry'
      $pagerProviderRegistry: '@fos_elastica.pager_provider_registry'

  FOS\ElasticaBundle\Message\Handler\:
    resource: '../vendor/friendsofsymfony/elastica-bundle/src/Message/Handler'
    autoconfigure: false
    tags:
     - name: messenger.message_handler
       handles: FOS\ElasticaBundle\Message\AsyncPersistPage
```

Now it does show when debugging the messenger:
```
Messenger
=========

command.bus
-----------

 The following messages can be dispatched:

 ----------------------------------------------------------------------------------                                                                                    
  FOS\ElasticaBundle\Message\AsyncPersistPage                                       
      handled by FOS\ElasticaBundle\Message\Handler\AsyncPersistPageHandler         
                                                                                    
  Symfony\Component\Mailer\Messenger\SendEmailMessage                               
      handled by mailer.messenger.message_handler                                   
                                                                                    
  Represents an update to send to the hub.                                          
  Symfony\Component\Mercure\Update                                                  
      handled by mercure.hub.default.publisher                                      
                                                                                    
 ---------------------------------------------------------------------------------- 
```

If I run the command to populate again, the situation is more strange:
- All the messages still fail directly with `No handler for message "FOS\ElasticaBundle\Message\AsyncPersistPage".
- Manually retrying with `php bin/console messenger:failed:retry 170` gives me `[OK] All done!`. The message is handled successfully.

**Comment by darthf1** [comment]
Definitely an issue on my side; I reinstalled / reconfigured from scratch and now its working.

`fos_elastica.yaml`
```yaml
# Read the documentation: https://github.com/FriendsOfSymfony/FOSElasticaBundle/blob/master/doc/setup.md
fos_elastica:
  messenger: ~
  serializer: ~
  clients:
    default: { url: '%env(ELASTICSEARCH_URL)%' }
  indexes:
    myentity:
      use_alias: true
      index_name: myentity
      serializer:
        groups: [myEntity:read]
      persistence:
        driver: orm
        model: App\Entity\MyEntity
        provider: ~
        finder: ~
```
`services.yaml`
```yaml
  FOS\ElasticaBundle\Message\Handler\AsyncPersistPageHandler:
    autoconfigure: false
    arguments:
      $persister: '@fos_elastica.async_pager_persister'
    tags: [{ name: messenger.message_handler, bus: command.bus }]
```

`messenger.yaml`
```yaml
framework:
  messenger:
    default_bus: command.bus
    # Uncomment this (and the failed transport below) to send failed messages to this transport for later handling.
    failure_transport: failed
    transports:
      # https://symfony.com/doc/current/messenger.html#transport-configuration
      priority:
        dsn: "%env(MESSENGER_TRANSPORT_DSN)%?queue_name=priority"
        retry_strategy:
          max_retries: 3
          delay: 1000
          multiplier: 2
          max_delay: 0
      default:
        dsn: "%env(MESSENGER_TRANSPORT_DSN)%?queue_name=default"
        retry_strategy:
          max_retries: 3
          delay: 1000
          multiplier: 2
          max_delay: 0
      elasticsearch:
        dsn: "%env(MESSENGER_TRANSPORT_DSN)%?queue_name=elasticsearch"
        retry_strategy:
          max_retries: 3
          delay: 1000
          multiplier: 2
          max_delay: 0
      failed: '%env(MESSENGER_TRANSPORT_DSN)%?queue_name=failed'
      # sync: 'sync://'

    routing:
      # Route your messages to the transports
      "*": default
      'FOS\ElasticaBundle\Message\AsyncPersistPage': elasticsearch
      'Symfony\Component\Mailer\Messenger\SendEmailMessage': priority

    buses:
      command.bus: ~
```

`supervisor.conf`
```
[supervisord]
logfile=/var/log/supervisord.log
pidfile=/var/run/supervisord.pid
nodaemon=false

[program:messenger-priority-default]
command=/srv/api/bin/console messenger:consume priority default --time-limit=3600
user=root
numprocs=2
autostart=true
autorestart=true

[program:messenger-elasticsearch]
command=/srv/api/bin/console messenger:consume elasticsearch --limit=10 --time-limit=3600
user=root
numprocs=2
autostart=true
autorestart=true
```
